### PR TITLE
Check status code of API request, prepare for POST requests.

### PR DESCRIPTION
**Pull request by cresh**

### CODE REVIEW

**Review by neshtea — COMMENTED** [review]
Please ignore this and see the review comment.

File: lib/api.ml [hunk 1]
```diff
@@ -52,36 +60,47 @@ type 'a api_request =
...
-  >>= fun (_, body) ->
-  Cohttp_lwt.Body.to_string body
-  >|= fun body -> Yojson.Safe.from_string body |> json_body_fn
+  >>= fun (response, body) ->
+  let status_code = Cohttp.(Response.status response |> Code.code_of_status) in
+  let string_body = Cohttp_lwt.Body.to_string body in
+  if Cohttp.Code.is_success status_code then
+    (string_body >|= fun body -> Yojson.Safe.from_string body |> json_body_fn)
+  else
+    (string_body >|= fun body -> Decoder.Yojson.Safe.fail (Printf.sprintf "Status Code %d: %s" status_code body) (Yojson.Safe.from_string body))
```
Comment (neshtea) [inline lib/api.ml]: I think it's a good idea to harden this and deal with other response codes. However, I don't think that this should result in a JSON decoder error -- the actual failure lies somewhere else. Maybe the result should be of type result where the error part differentiates between the different kinds of errors (e.g. connection error, json decoder error, ...), something like:

```ocaml
type response_error =
  | Http_error of int
  | Json_decoder_error of Decoder.Yojson.Safe.Error.t
  | ...
```

File: lib/api.ml [hunk 2]
```diff
@@ -41,9 +41,17 @@ let request_headers api_user api_pwd =
     ]
 ;;
 
-type request_method = GET
+type request_method = | GET | POST
```
Comment (neshtea) [inline lib/api.ml]: I think I started this off wrong. Now that I see the second `request_method`, I don't think we want to talk about request-methods at all but construct different types of `Api_requests`. Something along the lines of 

```ocaml
type body

type 'a request_info =
  { endpoint : string
  ; request_args : request_args
  ; decoder : 'a Decoder.Yojson.Safe.decoder
  }

type 'a api_request =
  | Api_get_request of 'a request_info
  | Api_post_request of 'a request_info * body
```

That would also remove the optional body and make it explicit that `GET`-Requests never have a body.

**Review by neshtea — CHANGES_REQUESTED** [review]
Thanks for the pull request. I see why one would need this. There are two points I'd like to discuss (see review).

File: lib/api.ml (hunk 2 above)
Comment (cresh) [reply lib/api.ml]: Addressed this with [017e971](https://github.com/neshtea/kimai_report/pull/1/commits/017e971838ff86ab3ad87dcff90ad51cfeab48f7)

File: lib/api.ml (hunk 1 above)
Comment (cresh) [reply lib/api.ml]: Addressed this with [22fb949](https://github.com/neshtea/kimai_report/pull/1/commits/22fb9498f8d91c87ebb729795a23683d8c9726c1)

**Pull request merged**